### PR TITLE
metal : single-source fusion table + fusion debug rework

### PR DESCRIPTION
**Pull request by ggerganov**
## Overview

Rework the Metal op fusion into a single source of truth: a fixed pattern table (`ggml_metal_fusions`) consumed by both the graph optimizer (packing) and the op encoders (compute), so the two phases can never disagree about which nodes fuse. Also add a fused `gated_delta_net + cpy` pattern that writes recurrent state snapshots directly into the KV cache, eliding the trailing cpy.

## Fusion regression test

Add a new test tool `test-fusion` that validates the fusion logic against a committed per-device CSV baseline. It runs every dummy model generated by `test-llama-archs` on a single device with fusion enabled and disabled, and reports the per-fusion-type counters plus the NMSE between the fused/unfused and device/CPU logits.

- The counters are gathered through an ad-hoc `get_proc_address` API. `ggml_backend_fusion_get` resolves the device fusion context once and returns an opaque `ggml_backend_fusion_t` handle; the stats functions (`ggml_backend_fusion_stats_init/reset/get/set_enabled`) operate on that handle instead of the backend device.
- The device owns a shared fusion context that every new backend context registers with, forcing single-threaded encoding while stats are collected, so the counters are race-free and accumulate across contexts.
- A fusion pattern that silently stops matching (or fires when it should not) is caught as a regression by comparing the counters against the committed `tests/fusion/MTL.csv` baseline.
- The unfused run doubles as a control: its counters must be all-zero.
- Prefill and decode usually produce the same per-graph count for the deterministic table, so the baseline stores a single `any` row per label (150 rows for Metal) instead of separate rows per mode.
- Both the count and the NMSE checks are hard gates. Two archs needed fixes to pass: plamo2 (graph fix) and qwen3tts (the codec-head `-inf` padding made the NMSE NaN; the dummy model now uses the exact codec-head vocab so the padding is not generated).
- The dummy models are generated in a separate CI step (`test_llama_archs_models`). The test itself runs gated on `GG_BUILD_METAL` atm, but can be extended in the future to other backends.

## Additional information

- `ggml_metal_fusion` gains an `unsafe` flag for fusions that are **not** elision chains (e.g. gdn + cache cpy). The generic chain/shape and `ggml_can_fuse_subgraph_ext` checks assume a linear element-wise chain with shape-preserving ops and intermediates consumed only within the pattern; the gdn+cpy fusion violates all three (the cpy reads a view of the gdn, shapes differ, and the gdn output feeds other consumers). For `unsafe` patterns those generic checks are skipped and the pattern-specific check callback becomes the sole validator, so it must re-establish the safety guarantees itself (op sequence, no graph output, exact snapshot-tail offset, cache layout, allocation).
- Packing (`ggml_metal_fuse_max`) now matches on the same view-transparent node sequence the compute phase uses, so fused groups (including intermediate views) stay adjacent through the reorder.
- The gated_delta_net kernel writes snapshots straight into the cache buffer (`nb_out` slot stride), so the trailing cpy is elided. (from https://github.com/ggml-org/llama.cpp/pull/25788)
- Verified on Metal: full `test-backend-ops` MTL0 suite passes (0 failures), including the new `GATED_DELTA_NET_CACHE_FUSION` tests; `test-fusion --check` passes (150 rows); Qwen3.5-0.8B Q8_0 shows no regression (~+5% tg from the fusion).

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES. pi:llama.cpp/DeepSeek-V4-Flash-Vision-Exp

### CODE REVIEW

File: ggml/src/ggml-metal/ggml-metal-device.m [hunk 1]
```diff
@@ -898,6 +898,9 @@ void ggml_metal_encoder_end_encoding(ggml_metal_encoder_t encoder) {
 
     // virtual address for GPU memory allocations
     atomic_uintptr_t addr_virt;
+
+    // shared fusion debugging context (NULL unless a test enabled fusion debugging)
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-device.m]:
```suggestion
    // shared fusion debugging context
```

File: ggml/src/ggml-metal/ggml-metal-ops.cpp [hunk 2]
```diff
@@ -33,22 +34,20 @@ struct ggml_metal_op {
         ggml_cgraph * gf,
         int  idx_start,
         int  idx_end,
-        bool use_fusion,
         bool use_concurrency,
         bool use_capture,
         int  debug_graph,
-        int  debug_fusion) {
+        ggml_metal_fusion * fusion) {
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-ops.cpp]: move up

File: ggml/src/ggml-metal/ggml-metal-fuse.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    // check callback below is the sole validator (used for patterns that are not elision chains,
+    // e.g. the gdn + cache-cpy write-through fusion)
+    bool unsafe;
+
+    // extra backend constraints on top of ggml_can_fuse_subgraph
+    // nodes[j] is the j-th node of the pattern
+    bool (*check)(const struct ggml_tensor * const     * nodes,
+                  const struct ggml_metal_fuse        * fuse,
+                        ggml_metal_fuse_mode            mode);
+};
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-fuse.h:58]: fix indent / valign

File: tests/fusion/MTL.csv [hunk 4]
```diff
@@ -0,0 +1,302 @@
+# test-fusion baseline for device MTL
+# arch              ,moe ,mode    ,label                       ,  count
+arcee               ,0   ,prefill ,RMS_NORM+MUL                ,      5
+arcee               ,0   ,decode  ,RMS_NORM+MUL                ,     80
+arctic              ,0   ,prefill ,RMS_NORM+MUL                ,      7
+arctic              ,0   ,decode  ,RMS_NORM+MUL                ,    112
+baichuan            ,0   ,prefill ,RMS_NORM+MUL                ,      5
+baichuan            ,0   ,decode  ,RMS_NORM+MUL                ,     80
```
Comment (ggerganov) [inline tests/fusion/MTL.csv]: dedup prefill/decode rows when the counts match (mode = "any")

File: ci/run.sh [hunk 5]
```diff
@@ -334,6 +334,34 @@ function gg_sum_test_llama_archs_tensor_split {
...
+        ./build-ci-release/bin/test-fusion --models build-ci-fusion-models --device MTL0 --check tests/fusion/MTL.csv 2>&1
+    fi
+
+    set +e
+}
+
+function gg_sum_test_fusion {
+    gg_printf '### %s\n\n' "${ci}"
+
+    gg_printf 'Runs test-fusion against the Metal baseline\n'
```
Comment (ggerganov) [inline ci/run.sh]:
```suggestion
    gg_printf 'Runs test-fusion against baseline\n'
```

File: ci/run.sh [hunk 6]
```diff
@@ -334,6 +334,34 @@ function gg_sum_test_llama_archs_tensor_split {
...
+function gg_run_test_fusion {
+    cd ${SRC}
+
+    set -e
+
+    if [ ! -z ${GG_BUILD_METAL} ]; then
+        rm -rf build-ci-fusion-models && mkdir -p build-ci-fusion-models
+
+        # generate the dummy models and run the fusion regression test
+        ./build-ci-release/bin/test-llama-archs -o build-ci-fusion-models 2>&1
```
Comment (ggerganov) [inline ci/run.sh]: move to a separate step (the models will be used by other tests too)

File: ggml/src/ggml-metal/ggml-metal-context.m [hunk 7]
```diff
@@ -36,15 +37,14 @@
...
     bool use_graph_optimize;
 
     int debug_graph;
-    int debug_fusion;
 
-    // how many times a given op was fused
-    uint64_t fuse_cnt[GGML_OP_COUNT];
+    // shared fusion debugging context (NULL unless a test enabled fusion debugging)
+    // when non-NULL the graph is always encoded by a single thread (n_cb == 0)
+    struct ggml_metal_fusion * fusion;
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-context.m]:
```suggestion
    struct ggml_metal_fusion * fusion;
```

File: ggml/src/ggml-metal/ggml-metal-context.m [hunk 8]
```diff
@@ -138,28 +138,42 @@ ggml_metal_t ggml_metal_init(ggml_metal_device_t dev) {
...
 
         if (getenv("GGML_METAL_GRAPH_OPTIMIZE_DISABLE") != NULL) {
             res->use_graph_optimize = false;
         }
 
-        memset(res->fuse_cnt, 0, sizeof(res->fuse_cnt));
+        // when fusion debugging is enabled the backend registers with the shared, device-owned
+        // fusion debugging context and forces single-threaded encoding (n_cb == 0) so the
+        // counters are race-free
+        res->fusion = ggml_metal_device_get_fusion(dev);
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-context.m]:
```suggestion
        res->fusion = ggml_metal_device_get_fusion(dev);
```

File: ggml/src/ggml-metal/ggml-metal-context.m [hunk 9]
```diff
@@ -138,28 +138,42 @@ ggml_metal_t ggml_metal_init(ggml_metal_device_t dev) {
...
+        if (res->fusion->stats) {
+            if (!res->fusion->labels_set) {
+                int n = 0;
+                const ggml_metal_fuse * all = ggml_metal_fuse_all(&n);
+
+                // labels point into the static fuse table; labels[]/counts[] are freed by the
+                // device when the debug context is destroyed
+                res->fusion->labels = calloc(n > 0 ? n : 1, sizeof(char *));
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-context.m]:
```suggestion
                res->fusion->labels = calloc(n > 0 ? n : 1, sizeof(char *));
```

File: ggml/src/ggml-metal/ggml-metal-device.h [hunk 10]
```diff
@@ -320,6 +320,29 @@ bool ggml_metal_device_supports_op(ggml_metal_device_t dev, const struct ggml_te
 
 const struct ggml_metal_device_props * ggml_metal_device_get_props(ggml_metal_device_t dev);
 
+// shared fusion debugging context, owned by the device; newly created backend contexts for that
+// device register with it so the fusion counters are race-free and accumulate across contexts.
+// labels[] points into the static fuse table; labels[]/counts[] are allocated with calloc and
+// freed when the context is destroyed.
+struct ggml_metal_fusion {
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal-device.h:330]:
```suggestion
// shared fusion debugging context, owned by the device; newly created backend contexts for that
// device register with it so the fusion counters accumulate across contexts.
struct ggml_metal_fusion {
```
